### DISCUSSION TITLE
strongswan to load balance between 64 logical cores ?

### DISCUSSION THREAD

**Discussion by edvinaskairys** [strongswan/strongswan#2503] [Q&A] [answered]
Hello,

Maybe there're something that could be tuned for strongswan to utilize more than few logical cores if i've a server with 64 logical cores. I know it's somehow dependent on IPSec connection count. But my problem is that i've only about ~10 connections, but the CPU cores count is 64. And only about ~10 of them are utilized.

The CPU i've is:

```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                64
On-line CPU(s) list:   0-63
Thread(s) per core:    2
Core(s) per socket:    32
Socket(s):             1
NUMA node(s):          1
Vendor ID:             AuthenticAMD
CPU family:            25
Model:                 1
Model name:            AMD EPYC 7543P 32-Core Processor
Stepping:              1
CPU MHz:               2800.000
CPU max MHz:           2800.0000
CPU min MHz:           1500.0000
BogoMIPS:              5589.78
Virtualization:        AMD-V
L1d cache:             32K
L1i cache:             32K
L2 cache:              512K
L3 cache:              32768K
NUMA node0 CPU(s):     0-63
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc art rep_good nopl nonstop_tsc extd_apicid aperfmperf eagerfpu pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_l2 cpb cat_l3 cdp_l3 invpcid_single hw_pstate sme ssbd rsb_ctxsw ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq overflow_recov succor smca
``` 
I've google that there're some setting like **threads** in **swanctl**, but i'm not sure if this would help.

Maybe there're other ways to make Strongswan utilize more system cores ? Currently i'm on old Centos 7.9, but soom we will be migrating to Rocky 9.

Thanks

**Comment by tobiasbrunner** [comment, marked as answer]
> I've google that there're some setting like **threads** in **swanctl**, but i'm not sure if this would help.

The ESP traffic is not handled by strongSwan, so it won't matter how many threads the IKE daemon (and especially not the `swanctl` config/control utility) uses.


> But my problem is that i've only about ~10 connections, but the CPU cores count is 64. And only about ~10 of them are utilized.

That's actually quite a nice spread for 10 SAs. Using multiple cores per SA is currently not really possible with the Linux kernel (there was the old [pcrypt](https://wiki.strongswan.org/projects/strongswan/wiki/pcrypt) stuff, but I think that has been pretty much abandoned).  You could look into the `per-cpu-sas-poc` branch (PoC-level implementation of [RFC 9611](https://datatracker.ietf.org/doc/html/rfc9611)), but the corresponding changes for the Linux kernel are also not yet merged (you'll find the [`xfrm-pcpu-v*` branches](https://git.kernel.org/pub/scm/linux/kernel/git/klassert/linux-stk.git/refs/) in Steffen Klassert's kernel repository). And it obviously has to be supported by both ends.

**Reply by edvinaskairys** [reply]
as always - thank you very much !
